### PR TITLE
Data-driven `line-dasharray` and `line-cap`

### LINKED ISSUE #3045: Implement property functions for "line-dasharray" and "line-cap"
Implementing property functions for `line-dasharray` will require sharing information about `LineAtlas` with the worker, much like the `*-pattern` properties https://github.com/mapbox/mapbox-gl-js/issues/2732

### PR DESCRIPTION
**Pull request by mourner**
Closes #3045. Implements data-driven expressions support for `line-dasharray` and `line-cap`. Example:

```js
"paint": {
  "line-dasharray": [
    "match", ["get", "property"],
    1, ["literal", [1, 2]],
    2, ["literal", [2, 2]],
    3, ["literal", [3, 2]],
    ["literal", [1, 1]]
  ]
},
"layout": {
  "line-cap": ["match", ["get", "property"], 2, "round", "butt"]
}
```

Similar to `imageAtlas` and `glyphAtlas`, `lineAtlas` is now generated per tile rather than globally, allowing for more dash values to fit the atlas, and moving dash SDF generation to the worker.

## Implementation status

- [x] Fix a bug where a zoom dasharray expression causes flickering when crossing the zoom stops zooming in.
- [x] Fix feature line-cap with constant dasharray
- [x] Fix feature line-cap with zoom dasharray
- [x] Fix composite line-cap with constant/zoom dasharray 
- [x] Trim line atlas to only the space used to avoid uploading 256x256px texture for every tile.

status | constant dash | zoom dash | feature dash | composite dash
--- | --- | --- | --- | ---
constant cap | 🟢  | 🟢 | 🟢 | 🟢
zoom cap  | 🟢 | 🟢 | 🟢 | 🟢
feature cap  | 🟢 | 🟢 | 🟢 | 🟢
composite cap  | 🟢 | 🟢 | 🟢 | 🟢

Something to follow up but not necessarily in this PR: we could switch to a statically packed sprite layout with `potpack`, similar to glyph and image atlases. This should drastically reduce the line atlas sizes while also improving rendering for more complex dashes, since the current dash SDF is fixed 256px x 1px or 15px (round) regardless of its configuration or expected line width range.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] post benchmark scores
 - [x] manually test the debug page
 - [x] tagged @mapbox/map-design-team @mapbox/static-apis if this PR includes style spec API or visual changes
 - [x] tagged @mapbox/gl-native if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Add support for data-driven expressions in line-dasharray and line-cap properties.</changelog>`

### CODE REVIEW

File: src/data/bucket/dash_attributes.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// @flow
+import {createLayout} from '../../util/struct_array.js';
+
+const dashAttributes = createLayout([
+    // [tl.x, tl.y, br.x, br.y]
+    {name: 'a_dash_to', components: 4, type: 'Uint16'},
+    {name: 'a_dash_from', components: 4, type: 'Uint16'},
+    {name: 'a_pixel_ratio_to', components: 1, type: 'Uint16'},
+    {name: 'a_pixel_ratio_from', components: 1, type: 'Uint16'},
```
Comment (ansis) [inline src/data/bucket/dash_attributes.js:9]: All these attributes aren't used, right?

Comment (mourner) [reply src/data/bucket/dash_attributes.js]: Yep, updated with a smaller layout.

**Comment by ansis** [comment]
This looks good to me. Can we add render tests to cover the all the possible variations?

**Comment by ansis** [comment]
<details>
<summary>Benchmark results</summary>

Screen Shot 2021-04-29 at 11 57 52 AM

</details>

After these changes load times are a bit more than 1% slower. It probably makes sense to merge these results anyways. @mourner can you comment to document why this is probably not something we can optimize away right now?

There is a slight bias in the results that we still need to fix but that only accounts for about 5ms of the difference.

**Comment by mourner** [comment]
> @mourner can you comment to document why this is probably not something we can optimize away right now?

The minor slowdown comes mostly from 2 things:
- Setting up and uploading a line atlas for every tile instead of a global one.
- Properties that may potentially be data-driven get treated differently in our program configuration code, adding overhead even if data-driven values aren't used.

The first item could potentially be addressed by keeping a separate global line atlas for non-data-driven dash layers, but this would add considerable complexity and isn't worth doing at the moment, at least until we reduced the overall complexity of dashed lines implementation by removing cross-fading #10605.

It's not clear how to address the second one without a major refactoring of how data-driven styling works, so we'll leave that for future optimizations.

**Review by ansis — APPROVED** [review]

File: src/shaders/line_sdf.vertex.glsl [hunk 2]
```diff
@@ -13,12 +13,11 @@ attribute float a_linesofar;
...
 uniform mediump float u_ratio;
 uniform lowp float u_device_pixel_ratio;
-uniform vec2 u_patternscale_a;
-uniform float u_tex_y_a;
-uniform vec2 u_patternscale_b;
-uniform float u_tex_y_b;
 uniform vec2 u_units_to_pixels;
 
+uniform vec2 u_texsize;
+uniform mediump vec3 u_scale;
```
Comment (astojilj) [inline src/shaders/line_sdf.vertex.glsl:19]: Could we change  `#define scale 0.015873016` (Line 7 here) to e.g. `SCALE` or, rename this uniform? It is related to metal shader compiling internals - u_scale gets transpiled to `u_.scale`, and there is name collision with constant `scale` `"error: '' :  syntax error, unexpected FLOATCONSTANT, expecting IDENTIFIER>"`.

**Comment by dythbj** [comment]
> Closes #3045. Implements data-driven expressions support for `line-dasharray` and `line-cap`. Example:
> 
> ```js
> "paint": {
>   "line-dasharray": [
>     "match", ["get", "property"],
>     1, ["literal", [1, 2]],
>     2, ["literal", [2, 2]],
>     3, ["literal", [3, 2]],
>     ["literal", [1, 1]]
>   ]
> },
> "layout": {
>   "line-cap": ["match", ["get", "property"], 2, "round", "butt"]
> }
> ```
> 
> Similar to `imageAtlas` and `glyphAtlas`, `lineAtlas` is now generated per tile rather than globally, allowing for more dash values to fit the atlas, and moving dash SDF generation to the worker.
> 
> ## Implementation status
> * [x]  Fix a bug where a zoom dasharray expression causes flickering when crossing the zoom stops zooming in.
> * [x]  Fix feature line-cap with constant dasharray
> * [x]  Fix feature line-cap with zoom dasharray
> * [x]  Fix composite line-cap with constant/zoom dasharray
> * [x]  Trim line atlas to only the space used to avoid uploading 256x256px texture for every tile.
> 
> status	constant dash	zoom dash	feature dash	composite dash
> constant cap	🟢	🟢	🟢	🟢
> zoom cap	🟢	🟢	🟢	🟢
> feature cap	🟢	🟢	🟢	🟢
> composite cap	🟢	🟢	🟢	🟢
> Something to follow up but not necessarily in this PR: we could switch to a statically packed sprite layout with `potpack`, similar to glyph and image atlases. This should drastically reduce the line atlas sizes while also improving rendering for more complex dashes, since the current dash SDF is fixed 256px x 1px or 15px (round) regardless of its configuration or expected line width range.
> 
> ## Launch Checklist
> * [x]  briefly describe the changes in this PR
> * [x]  write tests for all new functionality
> * [x]  document any changes to public APIs
> * [x]  post benchmark scores
> * [x]  manually test the debug page
> * [x]  tagged @mapbox/map-design-team @mapbox/static-apis if this PR includes style spec API or visual changes
> * [x]  tagged @mapbox/gl-native if this PR includes shader changes or needs a native port
> * [x]  apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
> * [x]  add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Add support for data-driven expressions in line-dasharray and line-cap properties.</changelog>`

@mourner Data-driven line-dasharray still not support on mapbox-gl 3.7.0?I have try this ,but got error blow:
image
'line-dasharray': [
          'match',
          ['get', 'nature'],
          ['case1', 'case2'],
          [
              "literal",
              [
                  2,
                  2,
                  6,
                  2
              ]
          ],
          [
              "literal",
              [
                  1,
                  1
              ]
          ]
      ]

Need you help please! @mourner

**Pull request merged**